Support for basic load/store in the interpreter

nbushehri opened the pull request.

nbushehri committed:
update gitignore

nbushehri committed:
Merge pull request #2 from yotann/master

Merge Sean's changes to alive-interpreter

nbushehri committed:
Merge branch 'bcdb_interp' of https://github.com/nbushehri/alive2 into bcdb_interp

nbushehri committed:
add missing llvm component required for building alive-worker-test

nbushehri committed:
Merge pull request #3 from yotann/master

Merge additional tests from yotann/alive2

nbushehri committed:
Add support for basic load/store operations in the Alive interpreter
This commit adds initial support for the i8 type, laying the groundwork for adding support for other bit widths.

nbushehri committed:
address code formatting issues from the previous commit

nbushehri committed:
Add support for load/store of more integer bitwidths
Also make interpreter report UB if out of bound memory access occurs.

yotann commented on util/worker.cpp:
```diff
@@ -427,6 +429,71 @@ ConcreteVal *WorkerInterpreter::loadConcreteVal(const IR::Type &type,
+ConcreteBlock WorkerInterpreter::loadConcreteBlock(const ojson &block, bool init) {
+  ConcreteBlock c_block;
+  c_block.size = block["size"].as_integer<uint64_t>();
+  c_block.address = 0; // TODO where would this be used
```
`c_block.address` will be used when there are inttoptr and/or ptrtoint instructions, but those aren't a priority right now (no change necessary). (alive.tv only generates addresses when there is at least one inttoptr or ptrtoint instruction in the code.)

yotann commented on ir/instr.cpp:
```diff
@@ -3577,6 +3577,66 @@ unique_ptr<Instr> Load::dup(const string &suffix) const {
   return make_unique<Load>(getType(), getName() + suffix, *ptr, align);
 }
 
+util::ConcreteVal *loadIntVal(Interpreter &interpreter,
```
This function should be static.

yotann commented on ir/instr.cpp:
```diff
@@ -3627,6 +3687,68 @@ unique_ptr<Instr> Store::dup(const string &suffix) const {
   return make_unique<Store>(*ptr, *val, align);
 }
 
+void storeIntVal(Interpreter &interpreter, util::ConcreteValPointer *ptr,
```
This function should be static.

yotann commented on ir/instr.cpp:
```diff
@@ -3577,6 +3577,66 @@ unique_ptr<Instr> Load::dup(const string &suffix) const {
+util::ConcreteVal *loadIntVal(Interpreter &interpreter,
+                              util::ConcreteValPointer *ptr,
+                              unsigned int bitwidth) {
+  assert(bitwidth >= 8);
```
Don't use an assert for a check like this that will be triggered sometimes in practice. Use interpreter.setUnsupported() and return nullptr or a default value (like `res` below). The problem with asserts is that the error message is never sent to the server, so there's no way to figure out how often the assert is being triggered or why a particular job failed.

yotann commented on ir/instr.cpp:
```diff
@@ -3577,6 +3577,66 @@ unique_ptr<Instr> Load::dup(const string &suffix) const {
+    return shared_ptr<ConcreteVal>(
+        loadIntVal(interpreter, c_ptr, getType().bits()));
+    interpreter.setUnsupported("load a value unsupported type");
+    return nullptr;
```
Delete these lines, they're unreachable.

yotann commented on ir/instr.cpp:
```diff
@@ -3577,6 +3577,66 @@ unique_ptr<Instr> Load::dup(const string &suffix) const {
+std::shared_ptr<util::ConcreteVal>
+Load::concreteEval(Interpreter &interpreter) const {
+  auto ptr_I = interpreter.concrete_vals.find(ptr);
+  assert(ptr_I != interpreter.concrete_vals.end());
```
Asserts are okay here (no change necessary) because triggering them should be impossible (unless there's a bug).

yotann commented on util/worker.cpp:
```diff
@@ -525,6 +592,8 @@ static ojson evaluateAliveInterpret(const ojson &options, const ojson &src,
 
   WorkerInterpreter interpreter(test_input);
+  if (options.contains("memory"))
+    interpreter.loadMemory(options["memory"]);
```
The memory belongs in `test_input`, not `options`. The idea is that the `test_input` produced by alive.tv can be passed directly into `alive.interpret`. The tests will look like this:

```yaml
args:
  - {}
  - !ir | 
    define void @tgt(i8* dereferenceable(1) %x) {
      store i8 0, i8* %x
      ret void
    }
  - args: [[1,0]]
    memory:
      - size: 0
        align: 0
      - size: 3
        align: 32
        bytes:
          - [0, 255, 128]
          - [null, 0, 0]
```

yotann commented on util/interp.h:
```diff
@@ -57,8 +198,10 @@ class Interpreter {
   std::string unsupported_reason;
   ConcreteVal *return_value = nullptr;
+  std::vector<ConcreteBlock> init_mem_blocks;
+  std::vector<ConcreteBlock> mem_blocks;
```
`mem_blocks` doesn't seem to be modified at all, so you should remove it. (For outlining, I don't need to compare the initial state against the final state; I just need to get the final state.)

yotann commented on util/worker.cpp:
```diff
@@ -427,6 +429,71 @@ ConcreteVal *WorkerInterpreter::loadConcreteVal(const IR::Type &type,
+                                             ptr_value[0].as_integer<uint64_t>(),
+                                             ptr_value[1].as_integer<uint64_t>());
+      auto cur_ptr_byte = ConcreteByte(move(concrete_ptr));
+      cur_ptr_byte.pointer_byte_offset = ptr_value[2].as_integer<int32_t>();
```
Should be `int64_t`, not `int32_t`.

yotann commented on ir/instr.cpp:
```diff
@@ -3577,6 +3577,66 @@ unique_ptr<Instr> Load::dup(const string &suffix) const {
+                              util::ConcreteValPointer *ptr,
+                              unsigned int bitwidth) {
+  assert(bitwidth >= 8);
+  assert ((bitwidth % 8) == 0); // FIXME
```
Use interpreter.setUnsupported().

yotann commented on util/interp.h:
```diff
@@ -3,22 +3,139 @@
+  bool is_pointer;
+  // for now, for non pointer bytes, treat the entire byte as an i8 for poison
+  // purposes
+  std::variant<ConcreteValPointer, ConcreteValInt> value;
```
It seems really wasteful to use `ConcreteValInt` here, since the value is always 8 bits. You just need an `uint8_t` for the value and another `uint8_t` for the poison bits.

yotann commented on util/interp.h:
```diff
@@ -3,22 +3,139 @@
 class ConcreteVal;
 
+struct ConcreteByte {
+  bool is_pointer;
```
It's redundant to use this with the variant. You can use `std::holds_alternative` to check what's in the variant. Or, if you prefer, you can keep this variable and replace the variant with two separate variables.

yotann commented on util/interp.h:
```diff
@@ -3,22 +3,139 @@
+      ub = false;
+    
+    if (!bytes.contains(index)) {
+      auto new_byte = ConcreteByte();
```
`new_byte` should be copied from `default_byte`.

yotann commented on ir/instr.cpp:
```diff
@@ -3577,6 +3577,66 @@ unique_ptr<Instr> Load::dup(const string &suffix) const {
+      return nullptr;
+    }
+
+    if (cur_byte == cur_block.default_byte || cur_byte.intValue().isPoison()) {
```
It's impossible for `cur_byte == cur_block.default_byte` to happen, so delete that part.

yotann reviewed (changes requested):
Thanks for working on this, and also thanks for figuring out my test format and writing tests! There are some changes I'd like you to make before accepting this, but it's on the right track.

yotann commented on util/worker.cpp:
```diff
@@ -427,6 +429,71 @@ ConcreteVal *WorkerInterpreter::loadConcreteVal(const IR::Type &type,
+  
+  for (auto& json_byte : block["bytes"].array_range()) {
+    assert(json_byte.size() == 3 && "each byte contains 3 elements");
+    if (json_byte[0].is_null()) {
```
You should load this byte just like the other bytes. You can make the value byte / ptr byte code below create the ConcreteByte, without emplacing it yet. Then add code after that to set `c_block.default_byte` or call `c_block.bytes.emplace` depending on whether `json_byte[0].is_null()` is true.

nbushehri replied:
I think I mistakenly assumed that the default byte is always the poison byte based on the example test files.
i.e. [null, 0, 0]. to confirm, the default byte can be any byte value?

nbushehri committed:
address some of the issues brought up in the pull request

nbushehri committed:
Address most of the remaining issues mentioned in the pull request review
Only issue that needs addressing is the loading of default bytes in worker
which I want to make sure we're on the same page before fixing.

nbushehri commented:
I think the latest commit (740d52f4c44ebe4f726a326de2ce04da0cc5b60a) addresses the issues you mentioned. There is one issue about default bytes that I need some clarification on. First, my understanding was that the default byte implies that the remaining bytes of the block are uninitialized, thus can be treated as poison. Or does it imply that the remaining bytes are a constant value specified by the 3-tuple [null, nonpoison_mask, value]?

yotann commented on util/interp.h:
```diff
@@ -21,40 +21,38 @@ namespace util {
-  ConcreteByte(ConcreteValInt &&_value)
-      : value(std::move(_value)), pointer_byte_offset(0) {}
+  ConcreteByte(DataByteVal &&_value)
+      : byte_val(std::move(_value)), pointer_byte_offset(0) {}
```
Needs to set `is_pointer`.

yotann reviewed (changes requested):
Yes, the default byte sets a constant value. `[null, 0, 0]` means the block is uninitialized, `[null, 255, 0]` means the block is zero.

You fixed most of my comments so I marked them as resolved. Just look at the few things I left open.

nbushehri committed:
address the issue with default bytes in worker-interp

nbushehri commented:
I think 26e2e946494c652f54158db86d45d9f7acbc9bf9 resolves the remaining issues.
